### PR TITLE
logging: Add support for QT_MESSAGE_PATTERN env var

### PR DESCRIPTION
**Pull request by Holzhaus**
This allows customizing the output of the log messages. The default should look almost exactly as before (except that the message type is lowercase).

If you want to customize, you can do that by setting the `QT_MESSAGE_PATTERN` variable, e.g.:

```
export QT_MESSAGE_PATTERN="`echo -e "\033[32m%{time h:mm:ss.zzz}%{if-category}\033[32m %{type}:%{endif} %{if-debug}\033[34m%{type}%{endif}%{if-warning}\033[31m%{type}%{endif}%{if-critical}\033[31m%{type}%{endif}%{if-fatal}\033[31m%{type}%{endif}\033[36m %{function} \033[33m[{{threadname}}]\033[0m %{message}"`"
```

### CODE REVIEW

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:171]: `+ QChar('\n')`

**Comment by uklotzde** [comment]
Code looks much nicer than before.

**Comment by uklotzde** [comment]
I would like to see this in 2.3. Opinions?

**Comment by ronso0** [comment]
That would be great in 2.3 as I could ditch `rainbow` then which seems to swallow some lines of output for example when Mixxx crashes.

**Comment by Holzhaus** [comment]
By the way, you can also quickly check where a log message comes from without grepping the source by using `%{file}:%{line}`.

**Comment by Holzhaus** [comment]
I'm unsure about allocations though. But I guess log messages from the engine thread are a no-go due to stderr I/O anyway...

**Comment by Holzhaus** [comment]
> I would like to see this in 2.3. Opinions?

I just tried rebasing but there's a merge conflict so i'll wait until a decision is made.

**Comment by uklotzde** [comment]
Allocations are not an issue, although we should minimize them. Besides that logging involves locks and file I/O. Nothing that should be done from the real-time thread.

Ideally handleMessage() would be delegated to a separate thread that performs the formatting and file I/O without the need for additional mutexes.

**Comment by poelzi** [comment]
With this patch I can either have colored output on console and colorcodes in the log files, or no colors at all.

**Comment by poelzi** [comment]
My branch now also uses the message pattern, but only for the console as otherwise colored output would mess up the file logging.
https://github.com/mixxxdj/mixxx/pull/3197

**Comment by daschuer** [comment]
Do we have an idea how this affects the performance?

Can you compare a Mixxx startup with this patch and without?

I like to avoid the situation that logging will we disabled entire in future for performance reasons, making us blind in support cases.

**Comment by uklotzde** [comment]
Why do we need to measure the startup performance? The startup takes seconds while any performance impact is measured in microseconds or less. We will not see any difference for a bunch of log messages.

**Comment by daschuer** [comment]
> Why do we need to measure the startup time? The startup takes seconds while any performance impact is measured in microseconds or less. We will not see any difference for a bunch of log messages.

That was the easiest test that jumps in my mind. What else can we do to judge about a performance impact?

**Review by uklotzde — APPROVED** [review]

**Review by uklotzde — CHANGES_REQUESTED** [review]
AppVeyor disagrees:

```
/FdCMakeFiles\mixxx-lib.dir\mixxx-lib.pdb /FS -c ..\src\util\logging.cpp
..\src\util\logging.cpp(55): warning C4267: 'initializing': conversion from 'size_t' to 'int', possible loss of data
..\src\util\logging.cpp(55): warning C4267: 'initializing': conversion from 'size_t' to 'const int', possible loss of data
..\src\util\logging.cpp(105): warning C4267: '+=': conversion from 'size_t' to 'int', possible loss of data
..\src\util\logging.cpp(125): warning C4267: '+=': conversion from 'size_t' to 'int', possible loss of data
..\src\util\logging.cpp(137): warning C4267: '+=': conversion from 'size_t' to 'int', possible loss of data
..\src\util\logging.cpp(149): warning C4267: '+=': conversion from 'size_t' to 'int', possible loss of data
..\src\util\logging.cpp(155): warning C4267: '+=': conversion from 'size_t' to 'int', possible loss of data
..\src\util\logging.cpp(253): error C3861: 'setenv': identifier not found
[560/886] Building CXX object CMakeFiles\mixxx-lib.dir\src\widget\weffectparameterknobcomposed.cpp.obj
```

**Comment by uklotzde** [comment]
Once performance becomes a real issue then logging has to be moved into a separate `MessageLogWorker` thread. I could take a look after this PR has been merged. Should be a straightforward extension.

Replacing the legacy message composition code is the right decision.

**Comment by daschuer** [comment]
> Once performance becomes a real issue then logging has to be moved into a separate MessageLogWorker thread. I could take a look after this PR has been merged. Should be a straightforward extension.

I am afraid that does not really work. In case of debug assertions, we will never see the log-output, because the process is stopped once the message was written in the message queue of the MessageLogWorker thread.

**Comment by uklotzde** [comment]
> > Once performance becomes a real issue then logging has to be moved into a separate MessageLogWorker thread. I could take a look after this PR has been merged. Should be a straightforward extension.
> 
> I am afraid that does not really work. In case of debug assertions, we will never see the log-output, because the process is stopped once the message was written in the message queue of the MessageLogWorker thread.

We can add a separate, immediate output for level QtCriticalMsg and higher. In these cases, a performance impact doesn't matter.

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:85]: The locking could be moved down 2 lines

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:111]: Writing to stderr from different threads will cause interleaved log messages

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:110]: `WriteFlags flags` is redundant and confusing, we only need a boolean flush flag here.

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:83]: Same here, replace flags with a boolean

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:246]: Why WriteFlag::All instead of writeFlags?

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/util/logging.cpp
Comment (Holzhaus) [reply src/util/logging.cpp:246]: because I'm stupid :D Fixed.

File: src/util/logging.cpp
Comment (Holzhaus) [reply src/util/logging.cpp]: One more reason to use a message queue. But that issue wasn't introduced here. Do you want to look into this?

Comment (uklotzde) [reply src/util/logging.cpp]: We need to use an additional mutex until then. Or use the file mutex for both as before.

Comment (Holzhaus) [reply src/util/logging.cpp]: https://github.com/mixxxdj/mixxx/pull/3204/commits/df034bd217719f1405566c97c559954f2c324307

**Comment by poelzi** [comment]
Is the mutex allocated from the RT Thread if the engine prints ?

**Comment by Holzhaus** [comment]
> Is the mutex allocated from the RT Thread if the engine prints ?

The mutex is static, the locking happens in the RT thread. But in that case the file I/O will be the bigger problem I think. Maybe we can switch to a lock-free queue approach with a separate thread that write the output in a later PR.

**Comment by uklotzde** [comment]
> > Is the mutex allocated from the RT Thread if the engine prints ?
> 
> The mutex is static, the locking happens in the RT thread. But in that case the file I/O will be the bigger problem I think. Maybe we can switch to a lock-free queue approach with a separate thread that write the output in a later PR.

Logging from the RT thread is strictly forbidden because of the string memory allocations. A lock-free queue doesn't help.

File: src/util/logging.cpp
Comment (daschuer) [reply src/util/logging.cpp:312]: {{threadname}} should become a const variable

**Review by daschuer — CHANGES_REQUESTED** [review]
Here some comments. 
I have digged into the QT code and the implementation is quite smart. The pattern is evaluated only once so the performance penalty should be negligible.

**Comment by daschuer** [comment]
> Logging from the RT thread is strictly forbidden because of the string memory allocations. A lock-free queue doesn't help.

If we crash anyway, it is OK. Also in one time cases where a we have hick-up anyway. I think the current situation is OK no need regarding logging from RT threads.

File: src/util/logging.cpp
Comment (Holzhaus) [reply src/util/logging.cpp]: Have a look at https://github.com/mixxxdj/mixxx/pull/3204/commits/938c96281c904a9f8813c0f5af4acf76d4c84daa. I think we need a `reinterpret_cast` and should use `size_t` instead to not make assumptions about the pointer size.

Comment (daschuer) [reply src/util/logging.cpp]: I have picked qlonglong to explicit select the correct overload that works for all sizes. But if your version works as well, I don't really mind.

**Review by daschuer — APPROVED** [review]
LGTM, Thank you. Works also good.

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:148]: QTextStream (constructed with a QString pointer for the underlying buffer) should work:
https://doc.qt.io/qt-5/qtextstream.html#operator-lt-lt-18

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:24]: Add a newline and a separate comment

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/util/logging.cpp
Comment (uklotzde) [inline src/util/logging.cpp:314]: kDefaultMessagePattern? Then the comment becomes obsolete.

**Comment by uklotzde** [comment]
main or 2.3?

**Comment by Holzhaus** [comment]
> main or 2.3?

I'm not particularily keen on resolving all the merge conflicts when rebasing on 2.3 tbh.

**Comment by uklotzde** [comment]
> > main or 2.3?
> 
> I'm not particularily keen on resolving all the merge conflicts when rebasing on 2.3 tbh.

Agreed.

**Comment by uklotzde** [comment]
LGTM. Awaiting CI results.

**Review by uklotzde — APPROVED** [review]

**Review by uklotzde — CHANGES_REQUESTED** [review]
Code style issue: A strayed, empty line

**Comment by uklotzde** [comment]
Remaining CI test failures are unrelated and will be fixed by #3209. LGTM

**Review by uklotzde — APPROVED** [review]

**Pull request merged**
